samza-test/src/test/java/org/apache/samza/test/processor/TestZkStreamProcessor.java

```diff
@@ -0,0 +1,558 @@
+    try {
+      startCountDownLatch1.await(1000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      Assert.fail("got interrupted while waiting for the first processor to start.");
+    }
+
+    // make sure it consumes all the messages from the first batch
+    int attempts = 5;
+    while (attempts > 0) {
+      long leftEventsCount = TestStreamTask.endLatch.getCount();
+      System.out.println("current count = " + leftEventsCount);
+      if (leftEventsCount == totalEventsToGenerate - messageCount) { // read first batch
+        System.out.println("read all. current count = " + leftEventsCount);
+        break;
+      }
+      TestZkUtils.sleepMs(1000);
+      attempts--;
+    }
+    Assert.assertTrue("Didn't read all the events in the first batch in 5 attempts", attempts > 0);
+
+    // start the second processor
+    CountDownLatch countDownLatch2 = new CountDownLatch(1);
+    StreamProcessor sp2 = createStreamProcessor("2", map, countDownLatch2, null);
+    Thread t2 = runInThread(sp2, TestStreamTask.endLatch);
+    t2.start();
+
+    // wait until the processor reports that it has started
+    try {
+      countDownLatch2.await(1000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      Assert.fail("got interrupted while waiting for the 2nd processor to start.");
+    }
+
+    // wait for at least one full debounce time to let the system to publish and distribute the new job model
+    TestZkUtils.sleepMs(3000);
+
+    // produce the second batch of the messages, starting with 'messageCount'
+    produceMessages(messageCount, inputTopic, messageCount);
+
+    // wait until all the events are consumed
+    // make sure it consumes all the messages from the first batch
+    attempts = 5;
+    while (attempts > 0) {
+      long leftEventsCount = TestStreamTask.endLatch.getCount(); // how much is left to read
+      System.out.println("2current count = " + leftEventsCount);
+      if (leftEventsCount == 0) { // should read all of them
+        System.out.println("2read all. current count = " + leftEventsCount);
+        break;
+      }
+      TestZkUtils.sleepMs(1000);
+      attempts--;
+    }
+    Assert.assertTrue("Didn't read all the leftover events in 5 attempts", attempts > 0);
+
+    try {
+      synchronized (t1) {
+        t1.notify();
+      }
+      synchronized (t2) {
+        t2.notify();
+      }
+      t1.join(1000);
+      t2.join(1000);
+    } catch (InterruptedException e) {
+      Assert.fail("Failed to join finished threads:" + e.getLocalizedMessage());
+    }
+
+    // processor1 will read 40 events, and then processor1 and processor2 will read 80 events together,
+    // but the expected values are the same 0-79 - we should get each value one time.
+    // Meanwhile the number of events we gonna get is 80+40=120
+    Map<Integer, Boolean> expectedValues = new HashMap<>(2 * messageCount);
+    for (int i = 0; i < 2 * messageCount; i++) {
```

Comment:
I would move this line and the following two lines into verifyNumMessages(). And just put the expectedMessageCount as the second input for verifyNumMessages.